<a href="https://colab.research.google.com/github/1ga5/PFAD/blob/main/task_4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Анализ и обработка результатов

В анализе данных важно наглядное представление результатов, чтобы их мог оценить заказчик. Это бизнес, где на кону громадный трафик, серверные мощности, личное время многотысячной аудитории. И одновременно это исследование.

Как в настоящей экспериментальной науке, ответ непредсказуем. Его точность зависит от умелого владения статистическими методами и от качества исходных данных. Перед тем, как начинать считать, проверьте, грамотно ли ваши данные подготовлены.

# Группировка данных

Слово «анализ» означает разбор, рассмотрение с разных сторон. Анализ данных начинают с разделения их на группы по какому-нибудь признаку. Эта операция называется группировка данных. Она помогает изучить материал более подробно, чтобы затем перейти к поиску взаимосвязей между отдельными группами.


Группировка оправданна, если данные чётко делятся по значимому признаку, а полученные группы близки к теме задачи. Например, когда есть данные обо всех покупках в супермаркете, можно смело заниматься группировкой. Так можно установить время наплыва покупателей и решить проблему пиковых нагрузок. Или посчитать средний чек — обычно для магазинов это ключевая метрика.


Стадии группировки хорошо укладываются в словесную формулу **split-apply-combine**:

•	разделить, **split** — разбиение на группы по определённому критерию;

•	применить, **apply** — применение какого-либо метода к каждой группе в отдельности, например, подсчёт численности группы методом count() или суммирование вызовом sum();

•	объединить, **combine** — сведение результатов в новую структуру данных, в зависимости от условий разделения и выполнения метода это бывает DataFrame и Series.

На картинке изображен принцип **split-apply-combine** для таблицы с экзопланетами. Посмотрим, как вообще идут дела с поиском экзопланет. Сначала данные делят по группам, где каждая группа — это год. Потом метод **count()** подсчитывает численность каждой группы. В итоге получаем новую структуру данных с группами, где каждая содержит год и число открытых за этот год экзопланет.


В Рandas для группировки данных есть метод **groupby()**. Он принимает как аргумент название столбца, по которому нужно группировать. В случае с делением экзопланет по годам открытия:

**print(exoplanet.groupby('discovered'))**

**<pandas.core.groupby.DataFrameGroupBy object at 0x7fc1e1ca3400>**


Применение метода **groupby()** к объекту типа DataFrame приводит к созданию объекта особого типа — **DataFrameGroupBy**. Это сгруппированные данные. Если применить к ним какой-нибудь метод Pandas, они станут новой структурой данных типа **DataFrame** или **Series**.
Подсчитаем сгруппированные по годам экзопланеты методом **count()**:

Если нужно сравнить наблюдения по одному показателю, метод применяют к **DataFrameGroupBy** с указанием на один столбец. Нас в первую очередь интересует радиус экзопланет: мы ищем другую Землю. Давайте получим таблицу с единственным столбцом 'radius':

**exo_number = exoplanet.groupby('discovered')['radius'].count()**

**print(exo_number)**

Получили Series, где по годам открытия расписано количество экзопланет, для которых удалось установить радиус.

Посмотрим, как меняется средний радиус открытых экзопланет год от года. Для этого надо сложить радиусы планет, открытых за определённый год, и поделить на их количество (которое мы уже нашли).

Сумма радиусов считается методом **sum()**:

**exo_radius_sum = exoplanet.groupby('discovered')['radius'].sum()**


**print(exo_radius_sum)**



## Задача
Вашему вниманию предлагается датасет с обзором показателей самоубийств с 1985 по 2016 год — Сравнение социально-экономической информации с показателями самоубийств по годам и странам.

Столбцами этой таблицы являются: страна, год, пол, возрастная группа, количество самоубийств, население, уровень самоубийств, сводный ключ по стране за год, ИЧР (индекс человеческого развития) за год, ВВП за год, ВВП на душу населения, поколение (на основе среднего по возрастной группе).

1. Загрузите таблицу master, содержащую данные о самоубийствах. Выведите на экран таблицу и оцените данные, обратите внимание на названия столбцов, переименуйте их, если это необходимо.

2. Убедитесь в том, что данные прошли предподготовку.  Пропущенные и неопределённые значения выявляет метод isna(), а суммарное количество таких значений — метод sum().

3. Сгруппируйте DataFrame по столбцу country, сохраните полученный результат в переменной country_grouping.

4. Посчитайте количество случаев самоубийств по странам, которые были совершены, методом **sum()**, указав, что выбираем один столбец suicides_no.

Сохраните результат в переменной country_sum и выведите первые 30 строк этой таблицы.



In [21]:
#Youre code is here
import pandas as pd

df = pd.read_csv("https://raw.githubusercontent.com/1ga5/PFAD/refs/heads/main/master.csv")

df

country  year     sex          age  suicides_no  population  \
0         Albania  1987    male  15-24 years           21      312900   
1         Albania  1987    male  35-54 years           16      308000   
2         Albania  1987  female  15-24 years           14      289700   
3         Albania  1987    male    75+ years            1       21800   
4         Albania  1987    male  25-34 years            9      274300   
...           ...   ...     ...          ...          ...         ...   
27815  Uzbekistan  2014  female  35-54 years          107     3620833   
27816  Uzbekistan  2014  female    75+ years            9      348465   
27817  Uzbekistan  2014    male   5-14 years           60     2762158   
27818  Uzbekistan  2014  female   5-14 years           44     2631600   
27819  Uzbekistan  2014  female  55-74 years           21     1438935   

       suicides/100k pop    country-year  HDI for year  gdp_for_year ($)   \
0                   6.71     Albania1987           NaN      2,156,624,900   
1                   5.19     Albania1987           NaN      2,156,624,900   
2                   4.83     Albania1987           NaN      2,156,624,900   
3                   4.59     Albania1987           NaN      2,156,624,900   
4                   3.28     Albania1987           NaN      2,156,624,900   
...                  ...             ...           ...                ...   
27815               2.96  Uzbekistan2014         0.675     63,067,077,179   
27816               2.58  Uzbekistan2014         0.675     63,067,077,179   
27817               2.17  Uzbekistan2014         0.675     63,067,077,179   
27818               1.67  Uzbekistan2014         0.675     63,067,077,179   
27819               1.46  Uzbekistan2014         0.675     63,067,077,179   

       gdp_per_capita ($)       generation  
0                     796     Generation X  
1                     796           Silent  
2                     796     Generation X  
3                     796  G.I. Generation  
4                     796          Boomers  
...                   ...              ...  
27815                2309     Generation X  
27816                2309           Silent  
27817                2309     Generation Z  
27818                2309     Generation Z  
27819                2309          Boomers  

[27820 rows x 12 columns]

In [22]:
print(df.columns)

df = df.set_axis(['country', 'year', 'sex', 'age_range', 'suicides_count', 'population',
       'suicides_per_100k_people', 'country-year', 'HDI',
       'gdp_for_year', 'gdp_per_capita', 'generation'], axis=1)

del df['country-year']

df

Index(['country', 'year', 'sex', 'age', 'suicides_no', 'population',
       'suicides/100k pop', 'country-year', 'HDI for year',
       ' gdp_for_year ($) ', 'gdp_per_capita ($)', 'generation'],
      dtype='object')


country  year     sex    age_range  suicides_count  population  \
0         Albania  1987    male  15-24 years              21      312900   
1         Albania  1987    male  35-54 years              16      308000   
2         Albania  1987  female  15-24 years              14      289700   
3         Albania  1987    male    75+ years               1       21800   
4         Albania  1987    male  25-34 years               9      274300   
...           ...   ...     ...          ...             ...         ...   
27815  Uzbekistan  2014  female  35-54 years             107     3620833   
27816  Uzbekistan  2014  female    75+ years               9      348465   
27817  Uzbekistan  2014    male   5-14 years              60     2762158   
27818  Uzbekistan  2014  female   5-14 years              44     2631600   
27819  Uzbekistan  2014  female  55-74 years              21     1438935   

       suicides_per_100k_people    HDI    gdp_for_year  gdp_per_capita  \
0                          6.71    NaN   2,156,624,900             796   
1                          5.19    NaN   2,156,624,900             796   
2                          4.83    NaN   2,156,624,900             796   
3                          4.59    NaN   2,156,624,900             796   
4                          3.28    NaN   2,156,624,900             796   
...                         ...    ...             ...             ...   
27815                      2.96  0.675  63,067,077,179            2309   
27816                      2.58  0.675  63,067,077,179            2309   
27817                      2.17  0.675  63,067,077,179            2309   
27818                      1.67  0.675  63,067,077,179            2309   
27819                      1.46  0.675  63,067,077,179            2309   

            generation  
0         Generation X  
1               Silent  
2         Generation X  
3      G.I. Generation  
4              Boomers  
...                ...  
27815     Generation X  
27816           Silent  
27817     Generation Z  
27818     Generation Z  
27819          Boomers  

[27820 rows x 11 columns]

In [23]:
print(df.isna().sum())
print(df.duplicated().sum())

country                         0
year                            0
sex                             0
age_range                       0
suicides_count                  0
population                      0
suicides_per_100k_people        0
HDI                         19456
gdp_for_year                    0
gdp_per_capita                  0
generation                      0
dtype: int64
0


In [24]:
# Среднее значение из каждой группы
df['HDI'] = df['HDI'].fillna(df.groupby('country')['HDI'].transform('mean'))

# Среднее значение по всей выборке
df['HDI'] = df['HDI'].fillna(df['HDI'].mean())

print(df.isna().sum())

country                     0
year                        0
sex                         0
age_range                   0
suicides_count              0
population                  0
suicides_per_100k_people    0
HDI                         0
gdp_for_year                0
gdp_per_capita              0
generation                  0
dtype: int64


In [26]:
country_grouping = df.groupby('country')

country_sum = country_grouping['suicides_count'].sum()

print(country_sum.head(30))

country
Albania                     1970
Antigua and Barbuda           11
Argentina                  82219
Armenia                     1905
Aruba                        101
Australia                  70111
Austria                    50073
Azerbaijan                  1656
Bahamas                       93
Bahrain                      463
Barbados                     177
Belarus                    59892
Belgium                    62761
Belize                       348
Bosnia and Herzegovina       318
Brazil                    226613
Bulgaria                   36388
Cabo Verde                    42
Canada                    107561
Chile                      40895
Colombia                   53080
Costa Rica                  6792
Croatia                    18429
Cuba                       41418
Cyprus                       412
Czech Republic             43687
Denmark                    15297
Dominica                       0
Ecuador                    20660
El Salvador                11683
Na

# Сортировка данных


Поиск необычного в группе — что среди планет, что среди меломанов — это прежде всего поиск чемпионов: объектов с выдающимися показателями по разным статьям. Как всю таблицу, так и отдельные группы изучают, сортируя строки по какому-либо столбцу.

В Pandas для этой операции есть метод **sort_values()**. У него два аргумента:

• **by = 'имя столбца'** — имя столбца, по которому нужно сортировать;

• **ascending:** по умолчанию True. Для сортировки по убыванию установите значение False.



Среди экзопланет интересны близкие по размерам к Земле. Есть ли такие? Отсортируем список по радиусу в порядке возрастания. Тогда в голове таблицы окажутся самые малые, на которых гравитация не прижмёт нас к полу.

**print(exoplanet.sort_values(by='radius').head(30))**

Оказывается, некоторые из уже открытых экзопланет по размерам близки не то что к Земле, но уже и к Луне! Получим список экзопланет с радиусом меньше земного. Смотрите, как логический оператор (здесь это <) задействуется в отборе элементов столбца. Дальше нам этот приём не раз понадобится.

**print(exoplanet[exoplanet['radius'] < 1])**

Но и этот список такой длинный, что изучать его лучше по частям. Экзопланеты, близкие по размерам к Земле, за последнее десятилетие открывали нередко. Можно изучать список открытых за каждый год. Например, для 2014 года (вновь обратите внимание на работу логического оператора, теперь это ==):

**print(exoplanet[exoplanet['discovered'] == 2014])**

А чтобы не тратить время на лишнее, поставим оба условия сразу. Для этого в Pandas есть логический оператор &, подобный оператору and языка Python. Напомним, его смысл на русском языке можно передать словами «и ещё»:

**exo_small_14 = exoplanet[ (exoplanet['radius']<1) & (exoplanet['discovered']==2014)]**

**print(exo_small_14)**

Отсортируем результат в порядке убывания радиуса.

**print(exo_small_14.sort_values(by = 'radius', ascending = False))**





## Задача
1. Выделим пятерки стран с самым низким и самым высоким уровнем самоубийств в мире.

In [30]:
#Youre code is here
print(country_sum.sort_values().head(5))

country
Dominica                  0
Saint Kitts and Nevis     0
San Marino                4
Antigua and Barbuda      11
Maldives                 20
Name: suicides_count, dtype: int64


In [34]:
#Youre code is here
print(country_sum.sort_values(ascending=False).head(5))

country
Russian Federation    1209742
United States         1034013
Japan                  806902
France                 329127
Ukraine                319950
Name: suicides_count, dtype: int64


2. Исследуйте распределение количества суидидов по гендерному признаку. Найдите лидеров и антилидеров по количеству самоубийств среди женщин. Повторяется ли картина по сравнению с предыдущим исследованием?

In [37]:
#Youre code is here
print(df[df['sex'] == 'female'].groupby('country')['suicides_count'].sum().sort_values(ascending=False))

country
Japan                    251630
Russian Federation       214330
United States            213797
France                    89419
Republic of Korea         82615
                          ...  
Antigua and Barbuda           1
Saint Kitts and Nevis         0
Oman                          0
San Marino                    0
Dominica                      0
Name: suicides_count, Length: 101, dtype: int64


In [38]:
print(df[df['sex'] == 'female'].groupby('country')['suicides_count'].sum().sort_values())


country
Oman                          0
San Marino                    0
Dominica                      0
Saint Kitts and Nevis         0
Antigua and Barbuda           1
                          ...  
Republic of Korea         82615
France                    89419
United States            213797
Russian Federation       214330
Japan                    251630
Name: suicides_count, Length: 101, dtype: int64


# Описательная статистика

Чтобы представить себе среднестатистического самоубийцу, понадобится описательная статистика.

Из её показателей для количественного описания данных в нашей задаче нужны четыре меры — максимум, минимум, медиана и среднее.

Наибольшее и наименьшее обычно вычисляют только по одному признаку. Например, можно получить минимальное и максимальное значение уровня самоубийств (количество самоубийств на 100 тыс.населения) (столбец 'suicides/100k pop'). Для поиска максимума вызывают метод max(), примененный к этому столбцу:

print(master['suicides/100k pop'].max())


In [40]:
#Youre code is here
print(df['suicides_per_100k_people'].max())

224.97


Самый высокий показатель оказался 224.97 самоубийств на 100 тыс.населения. Интересно, в какой стране такая ситуация.

Запросим из master строку с максимальным значением, прибегнув к логической индексации с условием

**master['suicides/100k pop'] == master['suicides/100k pop'].max()**

In [41]:
#Youre code is here
df[df['suicides_per_100k_people'] == df['suicides_per_100k_people'].max()]

country  year   sex  age_range  suicides_count  population  \
1258   Aruba  1995  male  75+ years               2         889   

      suicides_per_100k_people       HDI   gdp_for_year  gdp_per_capita  \
1258                    224.97  0.774271  1,320,670,391           17949   

           generation  
1258  G.I. Generation

Минимальное значение — ищут методом **min()**. Нам интересно установить случаи, когда уровень суицидов минимальный, но не нулевой.

Вот как мы создаём выборку  и находим в ней минимальное значение:

**master_drop_null = master[master['suicides/100k pop'] != 0]**

**print(master_drop_null['suicides/100k pop'].min())**


In [42]:
#Youre code is here
master_drop_null = df[df['suicides_per_100k_people'] != 0]
print(master_drop_null['suicides_per_100k_people'].min())

0.02


Получим название самой "счастливой" страны. Как и при поиске максимума, воспользуемся условием и логической индексацией

In [43]:
#Youre code is here
df[df['suicides_per_100k_people'] == master_drop_null['suicides_per_100k_people'].min()]

country  year     sex   age_range  suicides_count  population  \
23347  South Africa  2000  female  5-14 years               1     5079980   
23371  South Africa  2002  female  5-14 years               1     5081661   
23479  South Africa  2011    male  5-14 years               1     5198162   
25060      Thailand  1994  female  5-14 years               1     5609000   
25061      Thailand  1994    male  5-14 years               1     5821000   

       suicides_per_100k_people       HDI     gdp_for_year  gdp_per_capita  \
23347                      0.02  0.632000  136,361,854,808            3361   
23371                      0.02  0.646714  115,748,110,113            2769   
23479                      0.02  0.651000  416,878,162,441            8906   
25060                      0.02  0.666500  146,683,499,006            2784   
25061                      0.02  0.666500  146,683,499,006            2784   

         generation  
23347    Millenials  
23371    Millenials  
23479  Generation Z  
25060    Millenials  
25061    Millenials

Результат вывода — две страны, которые соответствуют минимальному значению.
На основе полученных данных можно сделать вывод, что число совершенных самоубийств на 100 тыс. населения  находится в диапазоне от 0,02 до 224.97 секунд, не включая пропущенные.

Это знание пригодится, чтобы разобраться с медианой и средним арифметическим.

Среднее и медиана оценивают значения в центре выборки. Если тех строк, где число с большим количеством самоубийц, столько же, сколько тех, где это число мало — среднее подойдет. Но когда есть оторванные от основной массы лидеры, их результаты сильно смещают значение среднего вверх. Вот почему оценивать в этом случае лучше медианой.


В геометрии медиана делит фигуру на две равные по площади. В статистике она делит выборку пополам: в одной половине значения меньше медианного, в другой больше. Логично, что для определения медианы список обязательно должен быть отсортирован — либо по возрастанию, либо по убыванию.
Когда количество значений нечётное, медиана будет равна тому значению, которое оказалось ровно посередине отсортированного набора. Если же количество данных чётное, то медиана рассчитывается как среднее арифметическое двух соседних чисел в середине набора.



В Pandas есть метод **median()**, который считает медиану. По аналогии с **min()** и **max()** его можно применять ко всей таблице, к отдельному столбцу или к сгруппированным данным. Теперь применим медиану ко всем значениям времени прослушивания в нашей таблице, исключив нулевые:



In [52]:
#Youre code is here
print(df[df['suicides_per_100k_people'] != 0]['suicides_per_100k_people'].sort_values().median())

8.26


Чтобы убедиться, что лидеры действительно смещают средний показатель вверх, найдём среднее арифметическое всех этих значений методом mean():

In [53]:
#Youre code is here
print(df[df['suicides_per_100k_people'] != 0]['suicides_per_100k_people'].mean())

15.14694039678831
